src/object/obj_rpc.h

```diff
@@ -327,6 +331,177 @@ CRT_RPC_DECLARE(obj_sync, DAOS_ISEQ_OBJ_SYNC, DAOS_OSEQ_OBJ_SYNC)
+	/* The object ID is used to elect leader for DTX recovery.
+	 * If it is empty, then it is for a readonly transaction.
+	 */
+	daos_unit_oid_t			 dcsh_leader_oid;
+	struct dtx_epoch		 dcsh_epoch;
+	struct dtx_memberships		*dcsh_mbs;
+};
+
+struct daos_cpd_ec_tgts {
+	uint32_t			 dcet_shard_idx;
+	uint32_t			 dcet_tgt_idx;
+};
+
+struct daos_cpd_update {
+	struct dcs_csum_info		*dcu_dkey_csum;
+	/* ID array for the DAOS targets that takes part in EC object update. */
+	struct obj_iod_array		*dcu_iod_array;
+	struct daos_cpd_ec_tgts		*dcu_ec_tgts;
+	/* Used for split EC update request. */
+	uint32_t			 dcu_start_shard;
+	/* see obj_rpc_flags. */
+	uint32_t			 dcu_flags;
+	union {
+		d_sg_list_t		*dcu_sgls;
+		crt_bulk_t		*dcu_bulks;
+	};
+};
+
+struct daos_cpd_punch {
+	daos_key_t			*dcp_akeys;
+};
+
+/**
+ * It is fake read style (daos fetch/enumerate/query) operation, only for set
+ * read timestamp on related target(s). It does not need iod/sgl information.
+ */
+struct daos_cpd_read {
+	daos_iod_t			*dcr_iods;
+};
+
+/**
+ * Each daos_cpd_sub_req stands for one simple DAOS operation that can be
+ * handled via single VOS API call.
+ */
+struct daos_cpd_sub_req {
+	/* See enum daos_cpd_sub_opc.*/
+	uint16_t			 dcsr_opc;
+	/* Size of 'dcu_ec_tgts', only used for updating of EC object. */
+	uint16_t			 dcsr_ec_tgt_nr;
+	uint32_t			 dcsr_nr;
+	union {
+		/* Used by CPD PRC and server side logic. */
+		daos_unit_oid_t		 dcsr_oid;
+		/* Used by client side cache. */
+		void			*dcsr_obj;
+	};
+	daos_key_t			 dcsr_dkey;
+	uint64_t			 dcsr_dkey_hash;
+	uint64_t			 dcsr_api_flags;
+	union {
+		struct daos_cpd_update	 dcsr_update;
+		struct daos_cpd_punch	 dcsr_punch;
+		struct daos_cpd_read	 dcsr_read;
+	};
+};
+
+/**
+ * Used for locating a sub request to be executed on the specified DAOS target.
+ */
+struct daos_cpd_req_idx {
+	/* Shard index of the object for the sub request on this DAOS target. */
+	uint32_t			 dcri_shard_idx;
+	/* The index (relative to the first sub request for its transaction)
+	 * of sub-request in the 'oci_sub_reqs' array. For parsing convenience,
+	 * DCSO_READ requests firstly, then modification ones. The update and
+	 * punch are sorted as their original executed order.
+	 */
+	uint32_t			 dcri_req_idx;
+};
+
+/**
+ * Each 'daos_cpd_disp_ent' describes all the sub requests that belong to
+ * one transaction (in spite of compounded or not) on one DAOS target.
+ */
+struct daos_cpd_disp_ent {
+	/* The count of read sub requests for the DTX on the DAOS target. */
+	uint32_t			 dcde_read_cnt;
+	/* The count of write sub requests for the DTX on the DAOS target.
+	 * It can be up to '2 ^ 16 - 1' at most because of the restriction
+	 * of 16-bits minor epoch on the server.
+	 */
+	uint32_t			 dcde_write_cnt;
+	/* Pointer to 'daos_cpd_req_idx' array. */
+	struct daos_cpd_req_idx		*dcde_reqs;
+};
+
+enum daos_cpd_sg_type {
+	DCST_HEAD	= 1,
+	DCST_REQ_CLI	= 2,
+	DCST_REQ_SRV	= 3,
+	DCST_DISP	= 4,
+	DCST_TGT	= 5,
```

Comment:
a few comments for the types are helpful, thx

Reply:
OK, I will add comment in subsequent patch.